components/speedreader/rust/lib/src/readability/src/dom.rs

```diff
@@ -190,3 +205,53 @@ pub fn text_children_count(handle: &Handle) -> usize {
     }
     count
 }
+
+pub fn previous_element_sibling<'a>(
+    handle: &Handle,
+    siblings: &'a Vec<Handle>,
+) -> Option<&'a Handle> {
+    let mut prev: Option<&Handle> = None;
+    for child in siblings.iter() {
+        if Rc::ptr_eq(handle, child) {
+            break;
+        }
+        if let Element { .. } = child.data {
+            prev = Some(child);
+        }
+    }
+    prev
+}
+
+pub fn parse_inner(contents: StrTendril) -> Option<Handle> {
+    let dom = parse_document(RcDom::default(), ParseOpts::default()).one(contents);
+    let document = dom.document.clone();
+    let html = document.children.borrow().get(0)?.clone();
+    let body = html.children.borrow().get(1)?.clone();
+    let img = body.children.borrow().get(0)?.clone();
```

Comment:
<details>
  <summary>Some notes about the cloning here, which looked problematic until I realized the <code>Handle</code> is actually an <code>Rc</code> type...</summary>
  You definitely don't need to `clone` so much here - I was able to remove the first one with no other changes; the rest will give you `temporary value dropped while borrowed` errors so you should be able to expand them like this:
  
  ```rust
  let document = dom.document;
  let document_children = document.children.borrow();
  let html = document_children.get(0)?;
  ```

  I believe you will be able to get away with just returning the `Option<&Handle>` from this function (which will require an explicit lifetime), but if not then just leave the last `clone` as-is.
</details>